### PR TITLE
MBS-8944 Add synthetic step for Entries which are out of steps

### PR DESCRIPTION
**Pull request by sergioboi**

### CODE REVIEW

File: subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt [hunk 1]
```diff
@@ -195,144 +166,84 @@ class ReportImplementation(
...
+            "Couldn't stop step because it hasn't started yet"
         }
 
-        val currentStep = currentState.currentStep
-        if (currentStep == null) {
-            throw RuntimeException("Couldn't stop precondition because it hasn't started yet")
-        }
+        currentState.testCaseStepList.add(currentStep)
+        afterStepStop(currentStep)
+        currentState.currentStep = null
```
Comment (sergioboi) [inline subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt:228]: Firstly I clear `currentStep` when it ends

File: subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt [hunk 2]
```diff
@@ -195,144 +166,84 @@ class ReportImplementation(
...
             throw RuntimeException("Couldn't stop precondition because it hasn't started yet")
         }
 
-        currentStep.apply {
-            currentState.preconditionStepList.add(this)
-            afterPreconditionStop(this)
-        }
+        currentState.preconditionStepList.add(currentStep)
+        afterPreconditionStop(currentStep)
+        currentState.currentStep = null
```
Comment (sergioboi) [inline subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt:207]: The same with precondition

File: subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt [hunk 3]
```diff
@@ -355,33 +266,24 @@ class ReportImplementation(
...
-
-            if (currentState is ReportState.Initialized.Started && currentState.currentStep != null) {
-                currentState.currentStep!!.futureUploads.add(it)
-            } else {
-                earlyFuturesUploads.add(it)
-            }
-        }
+        )
+        val started = getCastedStateOrNull<ReportState.Initialized.Started>()
+        val futureUploads = started?.getCurrentStepOrCreate {
```
Comment (sergioboi) [inline subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt:272]: And I get currentStep or create syntetic
But now I didn't call step callbacks. If I did for all entry would create a new synthetic step with a screenshot on each
If we need calling of those callbacks I will have to rewrite this logic

File: subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt [hunk 4]
```diff
@@ -393,28 +295,31 @@ class ReportImplementation(
         addEntry(Entry.Check(assertionMessage, timeProvider.nowInSeconds()))
     }
 
-    private data class Counted<T>(val t: T, var count: Int = 1)
+    private fun addEntry(entry: Entry) {
+        val started = getCastedStateOrNull<ReportState.Initialized.Started>()
+        val entriesList = started?.getCurrentStepOrCreate {
```
Comment (sergioboi) [inline subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt:295]: and there

File: subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt (hunk 3 above)
Comment (sergioboi) [inline subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt:272]: futureUploads are some kind of lazy steps

File: subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt [hunk 5]
```diff
@@ -393,28 +295,31 @@ class ReportImplementation(
...
-                acc + Counted(entry, 1)
-            }
-        }.map { counted: Counted<T> ->
-            if (counted.t is Entry.Comment && counted.count > 1) {
-                counted.t.copy(title = "[x${counted.count}] ${counted.t.title}")
-            } else {
-                counted.t
-            }
-        } as List<T>
+    private inline fun <reified T : ReportState> checkStateIs() = getCastedState<T>()
```
Comment (dsvoronin) [inline subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt:305]: these private functions looks like it belongs to ReportState class

**Review by dsvoronin — APPROVED** [review]

File: subprojects/common/report-viewer/src/main/kotlin/com/avito/report/model/Entry.kt [hunk 6]
```diff
@@ -1,16 +1,16 @@
 package com.avito.report.model
 
 sealed class Entry(
-    val type: String,
-    open val timeInSeconds: Long
+    val type: String
 ) {
+    abstract val timeInMs: Long
```
Comment (sergioboi) [inline subprojects/common/report-viewer/src/main/kotlin/com/avito/report/model/Entry.kt]: Fix bug when entries, which were at the same second, sorted incorrectly

**Pull request merged**